src/hotspot/share/utilities/concurrentHashTable.inline.hpp

```diff
@@ -1298,4 +1298,118 @@ inline bool ConcurrentHashTable<CONFIG, F>::
   return true;
 }
 
+template <typename CONFIG, MEMFLAGS F>
+class ConcurrentHashTable<CONFIG, F>::BucketsClaimer {
+  using InternalTable = ConcurrentHashTable<CONFIG, F>::InternalTable;
+
+  ConcurrentHashTable<CONFIG, F>* _cht;
+
+  struct InternalTableClaimer {
+    volatile size_t _next;
+    size_t _limit;
+    size_t _size;
+    InternalTable* _table;
+
+    InternalTableClaimer() : _next(0), _limit(0), _size(0), _table(nullptr) { }
+
+    InternalTableClaimer(size_t claim_size, InternalTable* table) :
+      InternalTableClaimer()
+    {
+      set(claim_size, table);
+    }
+
+    void set(size_t claim_size, InternalTable* table) {
+      assert(table != nullptr, "precondition");
+      _limit = table->_size;
+      _size  = MIN2(claim_size, _limit);
+      _table = table;
+    }
+  };
+
+  InternalTableClaimer _claimer;
+  // If there is a paused resize, we also need to operate on the already resized items.
+  InternalTableClaimer _new_table_claimer;
+public:
+  BucketsClaimer(ConcurrentHashTable<CONFIG, F>* cht, size_t claim_size) :
+    _cht(cht),
+    _claimer(claim_size, _cht->_table),
+    _new_table_claimer()
+  {
+    InternalTable* new_table = _cht->get_new_table();
+
+    if (new_table == nullptr) { return; }
+
+    DEBUG_ONLY(if (new_table == POISON_PTR) { return; })
+
+    _new_table_claimer.set(claim_size, new_table);
+  }
+
+  bool claim(InternalTableClaimer* claimer, size_t* start, size_t* stop, InternalTable** table) {
+    if (Atomic::load(&claimer->_next) < claimer->_limit) {
+      size_t claimed = Atomic::fetch_and_add(&claimer->_next, claimer->_size);
+      if (claimed < claimer->_limit) {
+        *start = claimed;
+        *stop  = MIN2(claimed + claimer->_size, claimer->_limit);
+        *table = claimer->_table;
+        return true;
+      }
+    }
+    return false;
+  }
```

Comment:
Minor nit: A better place for this method is probably `InternalTableClaimer`; then you can also probably improve encapsulation (visibility) of the members if you want, but I'm good with keeping everything public (as default for a struct) for this helper class.
I did not think through the suggestion earlier.

Reply:
you are right, i have updated, not certain about the `_table_claimer` and `_new_table_claimer` names, but couldn't come up with better